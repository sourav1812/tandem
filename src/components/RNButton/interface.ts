import {ReactNode} from 'react';
import {
  StyleProp,
  ViewStyle,
  PressableProps,
  TextStyle,
  LayoutChangeEvent,
} from 'react-native';

export interface Props {
  props?: PressableProps;
  customStyle?: StyleProp<ViewStyle>;
  onlyBorder?: boolean;
  color?: string;
  title?: string;
  buttonColor?: string;
  noBorderRadius?: boolean;
  onClick: () => Promise<void> | void;
  onlyIcon?: boolean;
  IconButtoncustomStyle?: StyleProp<ViewStyle>;
  icon?: ReactNode;
  textStyle?: StyleProp<TextStyle>;
  isDisabled?: boolean;
  ref?: any;
  onLayout?: ((event: LayoutChangeEvent) => void) | undefined;
  loadPercentage?: number;
  pressableStyle?: StyleProp<ViewStyle>;
  hitSlop?: number;
}
