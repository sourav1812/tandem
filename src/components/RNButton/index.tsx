/* eslint-disable react-native/no-inline-styles */
import {
  ActivityIndicator,
  Keyboard,
  Pressable,
  StyleSheet,
  View,
} from 'react-native';
import React from 'react';
import themeColor from '@tandem/theme/themeColor';
import RNTextComponent from '../RNTextComponent';
import {Props} from './interface';
import {verticalScale} from 'react-native-size-matters';
import {useAppSelector} from '@tandem/hooks/navigationHooks';
import {
  useSharedValue,
  withSequence,
  withTiming,
} from 'react-native-reanimated';
import Animated from 'react-native-reanimated';

const RNButton = ({
  props,
  customStyle,
  onlyBorder,
  title,
  buttonColor,
  noBorderRadius,
  onClick,
  onlyIcon,
  IconButtoncustomStyle,
  icon,
  textStyle,
  isDisabled = false,
  ref,
  loadPercentage,
  pressableStyle = {},
  onLayout,
  hitSlop,
}: Props) => {
  const isTablet = useAppSelector(state => state.deviceType.isTablet);
  const isButtonDisabled = useAppSelector(
    state => state.activityIndicator.isButtonDisabled,
  );
  const [pressed, setPressed] = React.useState(false);
  const disabled = isDisabled || isButtonDisabled;
  const scaleButton = useSharedValue(1);

  const runAnimation = () => {
    scaleButton.value = withSequence(
      withTiming(0.9, {duration: 200}),
      withTiming(1),
    );
  };

  const handlePress = async () => {
    Keyboard.dismiss();
    if (pressed) {
      return;
    }
    setPressed(true);
    await onClick();
    setPressed(false);
  };

  return (
    <>
      {!onlyIcon ? (
        <Pressable
          style={[pressableStyle && pressableStyle]}
          disabled={disabled}
          {...props}
          onLongPress={runAnimation}
          hitSlop={hitSlop}
          onPress={() => {
            if (disabled) {
              return;
            }
            if (handlePress) {
              runAnimation();
              handlePress();
            }
          }}>
          <Animated.View
            ref={ref && ref}
            onLayout={onLayout && onLayout}
            style={[
              styles.container,
              {transform: [{scale: scaleButton}]},
              {
                ...(buttonColor && {
                  borderColor: buttonColor,
                  backgroundColor: buttonColor,
                }),
              },
              {...(onlyBorder && {backgroundColor: 'white'})},
              {...(noBorderRadius && {borderRadius: 0})},
              isTablet && {maxHeight: verticalScale(35)},
              customStyle && customStyle,
            ]}>
            {isButtonDisabled && pressed ? (
              <ActivityIndicator
                style={{marginHorizontal: 15}}
                color={
                  onlyBorder
                    ? buttonColor || themeColor.themeBlue
                    : themeColor.white
                }
              />
            ) : (
              <RNTextComponent
                isSemiBold
                style={[
                  {
                    textAlign: 'center',
                    fontSize: verticalScale(14),
                    color: themeColor.white,
                    ...(onlyBorder && {
                      color: buttonColor || themeColor.themeBlue,
                    }),
                  },
                  textStyle && textStyle,
                ]}>
                {title}
              </RNTextComponent>
            )}
            {loadPercentage !== undefined && (
              <View
                style={{
                  width: `${100 - loadPercentage}%`,
                  height: '100%',
                  backgroundColor: '#00000088',
                  zIndex: -3,
                  position: 'absolute',
                  alignSelf: 'flex-end',
                }}
              />
            )}
          </Animated.View>
        </Pressable>
      ) : (
        <Pressable
          ref={ref && ref}
          onLayout={onLayout && onLayout}
          style={[
            styles.iconContainer,
            IconButtoncustomStyle && IconButtoncustomStyle,
          ]}
          {...props}
          onPress={handlePress}>
          {isButtonDisabled && pressed ? (
            <ActivityIndicator
              color={
                onlyBorder
                  ? buttonColor || themeColor.themeBlue
                  : themeColor.white
              }
            />
          ) : (
            icon
          )}
        </Pressable>
      )}
    </>
  );
};

export default RNButton;

const styles = StyleSheet.create({
  container: {
    borderWidth: 1,
    height: verticalScale(48),
    borderRadius: 14,
    justifyContent: 'center',
    alignItems: 'center',
    borderColor: themeColor.themeBlue,
    backgroundColor: themeColor.themeBlue,
  },
  iconContainer: {
    borderRadius: verticalScale(12),
    backgroundColor: '#F1F4F9',
    justifyContent: 'center',
    alignItems: 'center',
    height: verticalScale(36),
    width: verticalScale(36),
  },
});
