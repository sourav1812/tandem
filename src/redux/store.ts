import {AnyAction, combineReducers, configureStore} from '@reduxjs/toolkit';
import modeSlice from './slices/mode.slice';
import deviceTypeSlice from './slices/tablet.slice';
import orientationSlice from './slices/orientation.slice';
import activityIndicatorSlice from './slices/activityIndicator.slice';
import paramsReducer from './slices/paramsReducer';
import {MMKV} from 'react-native-mmkv';
import {persistReducer, Storage} from 'redux-persist';
import getResponseReducer from './slices/getResponseReducer';
import languageReducer from './slices/languageReducer';
import tokensSlice from './slices/tokens.slice';
import alertBoxSlice from './slices/alertBox.slice';
import storyGeneration from './slices/storyGeneration.slice';
import createChildSlice from './slices/createChild.slice';
import bookShelf from './slices/bookShelf.slice';
import setUserData from './slices/userData.slice';
import cache from './slices/cache.slice';
import setPermissions from './slices/permissions.slice';
import tooltipSlice from './slices/tooltip.slice';
import snapshotReducer from './slices/animationSnapshots.slice';
import storyLevelSlice from './slices/storyLevel.slice';
import recordingButtonSlice from './slices/recordingButton.slice';

export const storage = new MMKV();

export const appReducer = combineReducers({
  mode: modeSlice,
  deviceType: deviceTypeSlice,
  orientation: orientationSlice,
  activityIndicator: activityIndicatorSlice,
  params: paramsReducer,
  getResponseReducer: getResponseReducer,
  language: languageReducer,
  tokenReducer: tokensSlice,
  alertBoxReducer: alertBoxSlice,
  createChild: createChildSlice,
  userData: setUserData,
  permissions: setPermissions,
  storyGeneration,
  bookShelf,
  cache,
  tooltipReducer: tooltipSlice,
  snapshotReducer,
  storyLevel: storyLevelSlice,
  recording: recordingButtonSlice,
});

export const reduxStorage: Storage = {
  setItem: (key: string, value: string | number | boolean | Uint8Array) => {
    storage.set(key, value);
    return Promise.resolve(true);
  },
  getItem: (key: string) => {
    const value = storage.getString(key);
    return Promise.resolve(value);
  },
  removeItem: (key: string) => {
    storage.delete(key);
    return Promise.resolve();
  },
};

const persistConfig = {
  key: 'root',
  storage: reduxStorage,
};

export const rootReducer = (state: any, action: AnyAction) => {
  if (action.type === 'language/logoutFromRedux') {
    reduxStorage.removeItem('persist:root');
    storage.clearAll();
    return appReducer(undefined, action);
  }
  return appReducer(state, action);
};

const persistedReducer = persistReducer(persistConfig, rootReducer);

export const store = configureStore({
  reducer: persistedReducer,
  middleware: getDefaultMiddleware =>
    getDefaultMiddleware({
      serializableCheck: false,
      immutableCheck: false,
    }),
});

// Infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<typeof store.getState>;
// Inferred type: {posts: PostsState, comments: CommentsState, users: UsersState}
export type AppDispatch = typeof store.dispatch;
