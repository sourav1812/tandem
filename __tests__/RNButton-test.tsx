import RNButton from '@tandem/components/RNButton';
import React from 'react';
import 'react-native';
import renderer, {act} from 'react-test-renderer';
import {Provider} from 'react-redux'; // Import Provider
import configureStore from 'redux-mock-store'; // Import configureStore
import {Store, AnyAction} from '@reduxjs/toolkit';
import {Pressable} from 'react-native';
import {store} from '@tandem/redux/store';

// Create a mock Redux store
const mockStore = configureStore([]);

describe('RNButton Component', () => {
  let mstore: Store<unknown, AnyAction>;
  beforeEach(() => {
    mstore = mockStore(store.getState());
  });

  it('renders correctly', () => {
    const tree = renderer
      .create(
        <Provider store={mstore}>
          <RNButton onClick={() => {}} title="My Button" />
        </Provider>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('calls onClick when the RNButton is pressed', async () => {
    const onPressMock = jest.fn();
    const component = renderer.create(
      <Provider store={mstore}>
        <RNButton onClick={onPressMock} title="My Button" />
      </Provider>,
    );
    const button = component.root.findByType(Pressable);
    await act(() => {
      button.props.onPress();
    });

    expect(onPressMock).toHaveBeenCalled();
  });
});

// Writing test cases for a React Native app involves verifying that various components and functionalities work as expected. Below, I'll provide you with a set of test cases covering different aspects of testing for a React Native app. These are general guidelines, and you should adapt them to your specific app's features and requirements.

// ### Test Case 1: App Launch

// **Objective:** Ensure the app launches correctly and the initial screen is displayed.

// **Steps:**
// 1. Open the app.
// 2. Verify that the app's splash screen or loading screen is displayed.
// 3. Check that the app transitions to the main screen after loading.
// 4. Confirm that there are no crashes or errors during the launch.

// ### Test Case 2: Navigation

// **Objective:** Ensure that app navigation works correctly.

// **Steps:**
// 1. Navigate to various screens in the app using different navigation methods (e.g., stack navigation, tab navigation).
// 2. Check that the navigation bar or tabs correctly highlight the active screen.
// 3. Confirm that the back button or gesture navigation works as expected.
// 4. Verify that the app handles deep linking (if applicable).

// ### Test Case 3: User Authentication

// **Objective:** Validate the app's user authentication and registration processes.

// **Steps:**
// 1. Test user registration:
//    a. Fill out the registration form with valid data.
//    b. Confirm that the user is successfully registered.
//    c. Check for proper error handling with invalid data.
// 2. Test user login:
//    a. Enter valid login credentials.
//    b. Ensure the user is logged in and directed to the appropriate screen.
//    c. Attempt login with invalid credentials and verify appropriate error messages.
// 3. Test password reset functionality (if applicable):
//    a. Request a password reset email.
//    b. Confirm that the email is received.
//    c. Click on the reset link and set a new password.
//    d. Ensure the new password works for logging in.

// ### Test Case 4: Data Input and Validation

// **Objective:** Ensure data input forms and validation work correctly.

// **Steps:**
// 1. Test input validation:
//    a. Submit forms with empty fields and ensure appropriate validation errors are displayed.
//    b. Test different input validation scenarios (e.g., email format, password strength).
// 2. Test data submission:
//    a. Fill out various forms with valid data.
//    b. Confirm that the data is correctly submitted to the server or stored locally.
// 3. Test data retrieval:
//    a. Retrieve data from the server or local storage and display it in the app.
//    b. Ensure the displayed data matches the expected results.

// ### Test Case 5: UI/UX Testing

// **Objective:** Verify the user interface and user experience.

// **Steps:**
// 1. Test screen layouts:
//    a. Ensure that UI components are correctly positioned and aligned.
//    b. Verify that UI components are responsive to different screen sizes and orientations.
// 2. Test user interactions:
//    a. Check that buttons, links, and other interactive elements respond to user taps and gestures.
//    b. Confirm that any animations or transitions are smooth and without glitches.
// 3. Test color schemes and themes:
//    a. Ensure that the app's color scheme is consistent throughout.
//    b. Verify that dark and light mode (if supported) work correctly.

// ### Test Case 6: Network Connectivity

// **Objective:** Ensure the app functions correctly under different network conditions.

// **Steps:**
// 1. Test with a stable internet connection:
//    a. Verify that data is loaded from the server without issues.
//    b. Confirm that real-time features (e.g., chat, notifications) work as expected.
// 2. Test with limited or no internet connection:
//    a. Disable the internet connection and check for appropriate error handling.
//    b. Verify that the app gracefully handles offline scenarios and syncs data when the connection is restored.

// ### Test Case 7: Performance and Load Testing

// **Objective:** Evaluate the app's performance under different loads.

// **Steps:**
// 1. Test app startup time:
//    a. Measure how long it takes for the app to launch.
//    b. Ensure it loads within an acceptable time frame.
// 2. Test data loading time:
//    a. Load a large dataset and measure the time it takes to display.
//    b. Ensure that data loading is optimized for performance.
// 3. Test memory usage:
//    a. Monitor the app's memory consumption during normal usage.
//    b. Ensure the app does not have memory leaks or excessive memory usage.

// ### Test Case 8: Security Testing

// **Objective:** Check for potential security vulnerabilities.

// **Steps:**
// 1. Test for data validation:
//    a. Attempt to inject malicious code or scripts through input fields.
//    b. Ensure the app properly sanitizes and escapes user-generated content.
// 2. Test for unauthorized access:
//    a. Attempt to access restricted areas or features without proper authentication.
//    b. Verify that unauthorized access is denied.

// ### Test Case 9: Device Compatibility

// **Objective:** Ensure the app functions correctly on different devices and platforms.

// **Steps:**
// 1. Test on various Android devices:
//    a. Install and run the app on different Android devices with varying screen sizes and versions.
//    b. Ensure that the app layout and functionality adapt appropriately.
// 2. Test on various iOS devices:
//    a. Install and run the app on different iOS devices (iPhone and iPad) with varying screen sizes and iOS versions.
//    b. Confirm that the app works well on all tested iOS devices.

// ### Test Case 10: Accessibility Testing

// **Objective:** Verify that the app is accessible to users with disabilities.

// **Steps:**
// 1. Test screen reader compatibility:
//    a. Enable a screen reader (e.g., VoiceOver for iOS, TalkBack for Android).
//    b. Navigate through the app and ensure that all UI elements are properly labeled and accessible.
// 2. Test with high contrast mode:
//    a. Enable high contrast mode (if supported).
//    b. Verify that the app remains usable and readable.

// ### Test Case 11: Localization Testing

// **Objective:** Ensure the app works correctly with different languages and regions.

// **Steps:**
// 1. Change the device's language and region settings.
// 2. Verify that the app's text and content adapt to the selected language and region.

// Remember that these are general test cases, and the specific test cases you need will depend on your app's functionality and requirements. It's important to perform thorough testing throughout the development process and consider edge cases and user scenarios to ensure a high-quality React Native app. Additionally, consider using testing frameworks like Jest, React Testing Library, or Detox for automated testing where applicable.
