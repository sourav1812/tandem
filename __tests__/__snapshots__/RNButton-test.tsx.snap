// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RNButton Component renders correctly 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": false,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    [
      {
        "alignItems": "center",
        "backgroundColor": "#4285F6",
        "borderColor": "#4285F6",
        "borderRadius": 14,
        "borderWidth": 1,
        "height": 94.16470588235293,
        "justifyContent": "center",
      },
      {},
      {},
      {},
      false,
      undefined,
    ]
  }
>
  <Text
    style={
      [
        {
          "color": "black",
          "flexShrink": 1,
          "fontFamily": "Poppins-Regular",
          "fontSize": 30,
        },
        undefined,
        {
          "fontFamily": "Poppins-SemiBold",
          "fontSize": 16,
        },
        undefined,
        [
          {
            "color": "#ffffff",
            "fontSize": 27.46470588235294,
            "textAlign": "center",
          },
          undefined,
        ],
      ]
    }
  >
    My Button
  </Text>
</View>
`;
